File: Plugins/_flash_plugin_swift/flashplugin.swift
// Shared Flash plugin SDK for Swift (Foundation only) — no Flash business
// concepts, mirroring the Rust `flash_plugin` crate's role for Swift
// plugins. Scripts/build-plugins.sh compiles this file alongside each Swift
// plugin's main.swift.
//
// Speaks the wire contract from docs/plugin-protocol.md, whose constants
// are pinned by Plugins/_flash_plugin_specs/protocol.json: protocol v1, one
// JSON object per newline-terminated line over stdio, 10 MiB line cap both
// directions. Frame triage: id+method is a host request, id alone resolves
// a callHost waiter, method alone is a notification. The catalog is
// push-based (publish replaces it whole), perform is the single effect
// method with its four kinds routed to onResolve/onCommand/onAction/
// onNavigate, and stdin EOF is the shutdown signal. Lifecycle and ping
// answer synchronously on the read thread; handlers run on a worker queue
// so a slow AppleScript can never starve liveness.

import Foundation

// MARK: - Constants

/// The one wire protocol version this SDK speaks, echoed verbatim in every
/// initialize reply.
let protocolVersion = 1

/// One NDJSON line cap, both directions (protocol.json quotas.frame_bytes).
let maxFrameBytes = 10 << 20

/// Default callHost round-trip deadline.
let defaultCallTimeoutMs = 5000

// MARK: - Config / environment

/// Plugin configuration from the FLASH_PLUGIN_CONFIG env var, parsed once
/// (JSON object; empty when unset or invalid).
private let parsedConfig: [String: Any] = {
  guard let raw = ProcessInfo.processInfo.environment["FLASH_PLUGIN_CONFIG"],
    let data = raw.data(using: .utf8),
    let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
  else { return [:] }
  return object
}()

func config() -> [String: Any] { parsedConfig }

/// The host-provided writable data directory. Never defaults to "." — the
/// host always provides the dir, and failing loudly beats scattering plugin
/// state into an arbitrary working directory.
func dataDir() -> String {
  guard let dir = ProcessInfo.processInfo.environment["FLASH_PLUGIN_DATA_DIR"], !dir.isEmpty
  else {
    FileHandle.standardError.write(
      Data("flashplugin: FLASH_PLUGIN_DATA_DIR is unset — refusing to fall back to \".\"\n".utf8))
    exit(1)
  }
  return dir
}

/// JSONSerialization rejects Swift optionals; unwrap recursively
/// (nil → NSNull) so `[String: Any?]` payloads serialize.
private func jsonSafe(_ value: Any?) -> Any {
  guard let value else { return NSNull() }
  if let map = value as? [String: Any?] { return map.mapValues(jsonSafe) }
  if let array = value as? [Any?] { return array.map(jsonSafe) }
  return value
}

// MARK: - Runtime

final class PluginRuntime {
  // MARK: Framing

  private let output = FileHandle.standardOutput
  private let input = FileHandle.standardInput
  private let writeQueue = DispatchQueue(label: "flashplugin.write")  // the one serialized writer
  private let workers = DispatchQueue(label: "flashplugin.workers", attributes: .concurrent)

  /// Encode one frame onto stdout; false when encoding fails or the line
  /// would exceed the outbound cap (the frame is then not written at all —
  /// atomic rejection, never truncation).
  @discardableResult
  private func send(_ object: [String: Any?]) -> Bool {
    guard var frame = try? JSONSerialization.data(withJSONObject: jsonSafe(object)),
      frame.count <= maxFrameBytes
    else { return false }
    frame.append(0x0A)
    writeQueue.sync { output.write(frame) }
    return true
  }

  /// Method-only frame; an oversized notification is dropped.
  private func notify(_ method: String, _ params: [String: Any?]) {
    send(["method": method, "params": params])
  }

  /// The one reply an id'd request gets; a response over the outbound cap
  /// is replaced by the canonical frame-overflow error.
  private func respond(_ id: Int, _ result: [String: Any?]) {
    if !send(["id": id, "result": result]) {
      send(["id": id, "result": fail("response exceeded outbound frame limit")])
    }
  }

  // MARK: Pending / callHost

  private let pendingLock = NSLock()
  private var pending: [Int: ([String: Any?]) -> Void] = [:]
  private var nextRequestID = 1
  private var hostClosed = false

  /// Plugin→host RPC on our own id counter. Blocks the calling (worker)
  /// thread and never fails out-of-band: capability NAKs, host death
  /// ("host closed stdin"), and the deadline ("host call timed out") all
  /// arrive as ordinary {"ok": false, "error": …} results.
  func callHost(
    method: String, params: [String: Any?] = [:], timeoutMs: Int = defaultCallTimeoutMs
  ) -> [String: Any?] {
    pendingLock.lock()
    if hostClosed {
      pendingLock.unlock()
      return fail("host closed stdin")
    }
    let id = nextRequestID
    nextRequestID += 1
    let done = DispatchSemaphore(value: 0)
    var reply: [String: Any?] = [:]
    pending[id] = { result in
      reply = result
      done.signal()
    }
    pendingLock.unlock()
    send(["id": id, "method": method, "params": params])
    DispatchQueue.global(qos: .utility).asyncAfter(deadline: .now() + .milliseconds(timeoutMs)) {
      [self] in settle(id, fail("host call timed out"))  // removes pending: a late reply drops
    }
    done.wait()
    return reply
  }

  /// Resolve one pending call; responses to unknown ids drop silently.
  private func settle(_ id: Int, _ result: [String: Any?]) {
    pendingLock.lock()
    let completion = pending.removeValue(forKey: id)
    pendingLock.unlock()
    completion?(result)
  }

  // MARK: Dispatch

  private var initialized = false

  /// Frame triage: id+method → host request, id alone → callHost response,
  /// method alone → notification (unknown names are ignored). Lifecycle and
  /// ping answer right here on the read thread; handlers go to the workers.
  private func dispatch(_ message: [String: Any]) {
    let id = message["id"] as? Int
    let params = message["params"] as? [String: Any?] ?? [:]
    guard let method = message["method"] as? String else {
      if let id { settle(id, message["result"] as? [String: Any?] ?? fail("malformed host reply")) }
      return
    }
    guard let id else {
      if method == "event" {
        let name = params["name"] as? String ?? ""
        let payload = params["payload"] as? [String: Any?] ?? [:]
        workers.async { [self] in onEvent?(name, payload) }
      }
      return
    }
    switch method {
    case "initialize":
      guard !initialized else {
        respond(id, fail("initialize may only be called once"))
        return  // the one non-terminal protocol NAK: keep serving
      }
      let hostVersion = params["protocol_version"] as? Int ?? 0
      guard hostVersion == protocolVersion else {
        respond(
          id,
          [
            "ok": false, "protocol_version": protocolVersion,
            "error": "protocol version mismatch: host v\(hostVersion), plugin v\(protocolVersion)",
          ])
        exit(0)  // already flushed: writes are synchronous
      }
      initialized = true
      respond(id, ["ok": true, "protocol_version": protocolVersion])
      if let onStart { workers.async(execute: onStart) }  // AFTER the reply
    case "ping":
      respond(id, ["ok": true])
    case "evaluate":
      workers.async { [self] in
        respond(id, ["ok": true, "answers": (onEvaluate?(params) ?? []) as [Any?]])
      }
    case "search":
      workers.async { [self] in
        respond(id, ["ok": true, "rows": (onSearch?(params) ?? []) as [Any?]])
      }
    case "hints":
      workers.async { [self] in
        let (targets, contextPID) = onHints?(params) ?? ([], nil)
        var result: [String: Any?] = ["ok": true, "targets": targets as [Any?]]
        if let contextPID { result["context_pid"] = contextPID }
        respond(id, result)
      }
    case "perform":
      workers.async { [self] in respond(id, perform(params)) }
    default:
      respond(id, fail("unknown method: \(method)"))
    }
  }

  /// The single effect method's kind routing. An unregistered kind is "not
  /// mine" (the host may fall back); an unknown kind is an error (the host
  /// must not fall back on garbage).
  private func perform(_ params: [String: Any?]) -> [String: Any?] {
    let kind = params["kind"] as? String ?? ""
    let registered: ((_ params: [String: Any?]) -> [String: Any?])?
    switch kind {
    case "resolve": registered = onResolve
    case "command": registered = onCommand
    case "action": registered = onAction
    case "navigate": registered = onNavigate
    default: return fail("unknown perform kind: \(kind)")
    }
    guard let handler = registered else { return unhandled() }
    return handler(params)
  }

  // MARK: Handlers

  /// Runs on the worker queue AFTER the initialize reply (the reply is
  /// immediate by contract); typically ends with publish(rows).
  var onStart: (() -> Void)?
  /// Runs after stdin EOF, before the process exits 0.
  var onShutdown: (() -> Void)?
  var onEvent: ((_ name: String, _ payload: [String: Any?]) -> Void)?
  /// Synchronous, CPU-only answers for evaluate.
  var onEvaluate: ((_ params: [String: Any?]) -> [[String: Any?]])?
  /// Live rows in catalog row shape for search.
  var onSearch: ((_ params: [String: Any?]) -> [[String: Any?]])?
  /// Targets plus an optional context pid for hints.
  var onHints: ((_ params: [String: Any?]) -> ([[String: Any?]], Int?))?
  /// The four perform kinds; an unregistered kind answers the canonical
  /// {"ok": false, "unhandled": true}.
  var onResolve: ((_ params: [String: Any?]) -> [String: Any?])?
  var onCommand: ((_ params: [String: Any?]) -> [String: Any?])?
  var onAction: ((_ params: [String: Any?]) -> [String: Any?])?
  var onNavigate: ((_ params: [String: Any?]) -> [String: Any?])?

  // MARK: Emitters

  /// Replace the plugin's entire catalog (push-based; the host owns the
  /// store). Every row carries the first-class "source" field naming a
  /// manifest sources[].name; empty rows is an authoritative empty. On a
  /// transient refresh failure, simply don't publish — the host keeps the
  /// last-good catalog.
  func publish(_ rows: [[String: Any?]]) {
    notify("publish", ["rows": rows as [Any?]])
  }

  /// Feed manifest-declared status segments; "" clears one.
  func status(_ segments: [String: String]) {
    notify("status", ["segments": segments])
  }

  /// Structured, content-free logging: counts, stages, elapsed ms — never
  /// query text or candidate data.
  func log(_ level: String, _ message: String, fields: [String: String] = [:]) {
    notify("log", ["level": level, "message": message, "fields": fields])
  }

  // MARK: Serve loop

  /// Blocking read loop until stdin EOF — the shutdown signal: in-flight
  /// callHost waiters resolve to the canonical host-closed error, the
  /// onShutdown hook runs, and serve returns so main exits 0. An
  /// undecodable line is dropped; a line over the inbound cap is discarded
  /// through its next newline — the stream self-heals, never fatal.
  func serve() {
    var buffer = Data()
    var discarding = false
    while true {
      let chunk = input.availableData
      if chunk.isEmpty { break }  // stdin EOF
      buffer.append(chunk)
      while true {
        if discarding {
          guard let newline = buffer.firstIndex(of: 0x0A) else {
            buffer.removeAll(keepingCapacity: true)
            break
          }
          buffer.removeSubrange(buffer.startIndex...newline)
          discarding = false
        }
        guard let newline = buffer.firstIndex(of: 0x0A) else {
          if buffer.count > maxFrameBytes {  // oversized: discard to the next newline
            buffer.removeAll(keepingCapacity: true)
            discarding = true
          }
          break
        }
        let line = buffer.subdata(in: buffer.startIndex..<newline)
        buffer.removeSubrange(buffer.startIndex...newline)
        guard !line.isEmpty, line.count <= maxFrameBytes,
          let message = (try? JSONSerialization.jsonObject(with: line)) as? [String: Any]
        else { continue }
        dispatch(message)
      }
    }
    pendingLock.lock()
    hostClosed = true
    let inflight = pending
    pending = [:]
    pendingLock.unlock()
    for completion in inflight.values { completion(fail("host closed stdin")) }
    onShutdown?()
  }
}

// MARK: - Reply helpers

/// {"ok": true} carrying any extra reply fields.
func ok(_ fields: [String: Any?] = [:]) -> [String: Any?] {
  var result = fields
  result["ok"] = true
  return result
}

/// perform's "not my context" reply — the host MAY fall back.
func unhandled() -> [String: Any?] { ["ok": false, "unhandled": true] }

/// The ok:false error reply; keep messages content-free.
func fail(_ message: String) -> [String: Any?] { ["ok": false, "error": message] }

// MARK: - Subprocess helper

struct CommandResult {
  var ok: Bool
  var stdout: String
  var stderr: String
}

/// Bounded subprocess capture: hard timeout terminates the child, mirroring
/// the Rust SDK's run_command discipline (without the process-group kill —
/// osascript/open spawn no descendants worth chasing).
func runCommand(_ argv: [String], timeoutSeconds: Double) -> CommandResult {
  let process = Process()
  process.executableURL = URL(fileURLWithPath: argv[0])
  process.arguments = Array(argv.dropFirst())
  let out = Pipe()
  let err = Pipe()
  process.standardOutput = out
  process.standardError = err
  process.standardInput = FileHandle.nullDevice
  do {
    try process.run()
  } catch {
    return CommandResult(ok: false, stdout: "", stderr: "\(error)")
  }
  let killer = DispatchWorkItem { if process.isRunning { process.terminate() } }
  DispatchQueue.global(qos: .utility).asyncAfter(
    deadline: .now() + timeoutSeconds, execute: killer)
  let stdoutData = out.fileHandleForReading.readDataToEndOfFile()
  let stderrData = err.fileHandleForReading.readDataToEndOfFile()
  process.waitUntilExit()
  killer.cancel()
  return CommandResult(
    ok: process.terminationStatus == 0,
    stdout: String(data: stdoutData, encoding: .utf8) ?? "",
    stderr: String(data: stderrData, encoding: .utf8) ?? "")
}

func runOsascript(_ script: String, timeoutSeconds: Double) -> CommandResult {
  runCommand(["/usr/bin/osascript", "-e", script], timeoutSeconds: timeoutSeconds)
}

func applescriptQuote(_ value: String) -> String {
  "\""
    + value
    .replacingOccurrences(of: "\\", with: "\\\\")
    .replacingOccurrences(of: "\"", with: "\\\"")
    + "\""
}

File: Plugins/_flash_plugin_typescript/flashplugin.ts
// Shared Flash plugin SDK for TypeScript on Bun — no Flash business
// concepts, mirroring the Rust `flash_plugin` crate's role for TS plugins.
// Plugins import it by bare module name — the host (and the spec runner)
// inject NODE_PATH pointing at this directory at spawn, so the same import
// works from the checkout, the staged release bundle, and third-party roots:
//
//   import { Plugin, ok, unhandled, fail } from "flashplugin";
//
// Speaks the wire contract from docs/plugin-protocol.md (constants pinned by
// Plugins/_flash_plugin_specs/protocol.json): protocol v1, UTF-8 NDJSON over
// stdio, 10 MiB line cap both directions. Frame triage: id+method is a host
// request, id alone resolves a callHost pending, method alone is a
// notification. Registration is camelCase hooks on the constructor:
//
//   new Plugin({ onStart, onEvaluate, onCommand, ... }).serve();
//
// `perform` routes by kind to onResolve/onCommand/onAction/onNavigate; those
// hooks return replies built with ok(fields)/unhandled()/fail(msg) — plain
// object literals. onHints returns either a target array or {targets,
// context_pid} — the SDK wraps both as {ok: true, targets: [...]}. Request
// handlers are fired without awaiting so a hook that awaits callHost never
// stalls the read pump; all writes funnel through the one serialized writer
// (a single FileSink, one synchronous write+flush per frame). Hook errors
// never break the wire: request hooks answer fail("<method> hook failed")
// (evaluate answers empty — evaluators never error), lifecycle hooks log
// and continue. stdin EOF is the shutdown signal: onShutdown runs, serve()
// resolves, the process exits 0. callHost never rejects and never resolves
// undefined — timeouts and host death arrive as {ok: false, error} results.

// ── constants ──────────────────────────────────────────────────────────────

export const PROTOCOL_VERSION = 1;
const MAX_FRAME_BYTES = 10 * 1024 * 1024; // NDJSON line cap, both directions
const HOST_CALL_TIMEOUT_MS = 5000;

const ERR_FRAME_OVERFLOW = "response exceeded outbound frame limit";
const ERR_HOST_CLOSED = "host closed stdin";
const ERR_HOST_TIMEOUT = "host call timed out";

const PERFORM_KINDS = ["resolve", "command", "action", "navigate"] as const;

export type Value =
  | null
  | boolean
  | number
  | string
  | Value[]
  | { [key: string]: Value };

type Msg = { [key: string]: Value };

// Catalog row for publish/search: `source` is first-class and must name a
// manifest sources[].name.
export interface Row {
  source: string;
  title: string;
  url?: string;
  metadata?: { [key: string]: string };
  effect?: Value;
}

export interface Answer {
  title: string;
  subtitle?: string;
  effect: { type: string; text: string };
}

export type Reply = { [key: string]: Value };
export type HintTargets = Value[] | { targets: Value[]; context_pid?: number };

export interface Hooks {
  onStart?: () => void | Promise<void>;
  onShutdown?: () => void | Promise<void>;
  onEvent?: (name: string, payload: Value) => void | Promise<void>;
  onEvaluate?: (params: Msg) => Answer[]; // synchronous, CPU-only
  onSearch?: (params: Msg) => Row[] | Promise<Row[]>;
  onHints?: (params: Msg) => HintTargets | Promise<HintTargets>;
  onResolve?: (params: Msg) => Reply | Promise<Reply>;
  onCommand?: (params: Msg) => Reply | Promise<Reply>;
  onAction?: (params: Msg) => Reply | Promise<Reply>;
  onNavigate?: (params: Msg) => Reply | Promise<Reply>;
}

// ── reply helpers ──────────────────────────────────────────────────────────

export const ok = (fields: Reply = {}): Reply => ({ ok: true, ...fields });

export const unhandled = (): Reply => ({ ok: false, unhandled: true });

export const fail = (error: string): Reply => ({ ok: false, error });

// ── config / env accessors ─────────────────────────────────────────────────

let cachedConfig: { [key: string]: Value } | undefined;

/** The object parsed once from FLASH_PLUGIN_CONFIG ({} when unset/invalid). */
export function config(): { [key: string]: Value } {
  if (cachedConfig === undefined) {
    let parsed: Value = null;
    try {
      parsed = JSON.parse(process.env.FLASH_PLUGIN_CONFIG ?? "");
    } catch {}
    cachedConfig =
      parsed !== null && typeof parsed === "object" && !Array.isArray(parsed)
        ? parsed
        : {};
  }
  return cachedConfig;
}

/** The plugin's writable data directory; never defaults (a silent "." would
 * scatter state into whatever cwd the plugin happened to spawn in). */
export function dataDir(): string {
  const dir = process.env.FLASH_PLUGIN_DATA_DIR;
  if (!dir) throw new Error("FLASH_PLUGIN_DATA_DIR is not set");
  return dir;
}

/** Async single-process plugin runtime: one read pump, one serialized
 * writer, request hooks fired without blocking the pump. */
export class Plugin {
  private hooks: Hooks;
  private buf = "";
  private skipping = false; // inside an oversized inbound line
  private nextId = 0;
  private pending = new Map<number, (result: Reply) => void>();
  private initialized = false;
  private done = false; // stdin EOF or stdout death: cleanup, then exit 0
  private terminated = false; // version mismatch: exit 0 without cleanup
  // One writer for the process lifetime, flushed per frame: Bun's FileSink
  // buffers, and a fresh writer per send can lose the final reply when the
  // process exits before the sink drains.
  private writer = Bun.stdout.writer();

  constructor(hooks: Hooks = {}) {
    this.hooks = hooks;
  }

  // ── framing ────────────────────────────────────────────────────────────

  /** The single write path: one JSON object, one line, flushed. Returns
   * false when the encoded frame exceeds the outbound cap. */
  private send(obj: Value): boolean {
    const data = JSON.stringify(obj);
    if (Buffer.byteLength(data) > MAX_FRAME_BYTES) return false;
    try {
      this.writer.write(data + "\n");
      this.writer.flush();
    } catch {
      this.done = true; // host is gone; stdin EOF follows
    }
    return true;
  }

  private respond(id: Value, result: Reply) {
    if (!this.send({ id, result })) {
      this.send({ id, result: fail(ERR_FRAME_OVERFLOW) });
    }
  }

  private notify(method: string, params: Value) {
    this.send({ method, params }); // oversized notifications are dropped
  }

  // ── pending map / callHost ─────────────────────────────────────────────

  /** Plugin→host RPC. Never rejects and never resolves undefined — timeouts
   * and host death arrive as {ok: false, error} result objects. */
  callHost(
    method: string,
    params: Value = {},
    timeoutMs = HOST_CALL_TIMEOUT_MS,
  ): Promise<Reply> {
    const id = ++this.nextId;
    if (this.done) return Promise.resolve(fail(ERR_HOST_CLOSED));
    return new Promise((resolve) => {
      const timer = setTimeout(() => {
        this.pending.delete(id); // late replies are dropped
        resolve(fail(ERR_HOST_TIMEOUT));
      }, timeoutMs);
      this.pending.set(id, (result) => {
        clearTimeout(timer);
        resolve(result);
      });
      if (!this.send({ id, method, params })) {
        clearTimeout(timer);
        this.pending.delete(id);
        resolve(fail(ERR_FRAME_OVERFLOW));
      }
    });
  }

  private handleEof() {
    this.done = true;
    for (const [id, settle] of [...this.pending]) {
      this.pending.delete(id);
      settle(fail(ERR_HOST_CLOSED));
    }
  }

  // ── dispatch ───────────────────────────────────────────────────────────

  private dispatch(line: string) {
    let msg: Msg;
    try {
      msg = JSON.parse(line) as Msg;
    } catch {
      return; // wire noise is dropped, never fatal
    }
    if (msg === null || typeof msg !== "object" || Array.isArray(msg)) return;
    const method = typeof msg.method === "string" ? msg.method : undefined;
    const id = msg.id;
    if (method !== undefined && id !== undefined && id !== null) {
      // Host→plugin request: fired, not awaited — the pump keeps reading so
      // a hook that awaits callHost sees its reply arrive.
      void this.handleRequest(id, method, (msg.params ?? {}) as Msg);
    } else if (id !== undefined && id !== null) {
      // Response to one of our callHost requests; unknown ids are dropped.
      const settle = typeof id === "number" ? this.pending.get(id) : undefined;
      if (settle) {
        this.pending.delete(id as number);
        const result = msg.result;
        settle(
          result !== null && typeof result === "object" && !Array.isArray(result)
            ? (result as Reply)
            : fail("malformed host reply"),
        );
      }
    } else if (method === "event") {
      // Notification; unknown methods are ignored.
      const params = (msg.params ?? {}) as Msg;
      void (async () => this.hooks.onEvent?.(String(params.name ?? ""), params.payload ?? null))()
        .catch(() => this.log("error", "event hook failed"));
    }
  }

  // ── handler registry ───────────────────────────────────────────────────

  private async handleRequest(id: Value, method: string, params: Msg) {
    switch (method) {
      case "initialize":
        return this.handleInitialize(id, params);
      case "ping":
        return this.respond(id, ok());
      case "evaluate": {
        let answers: Answer[] = [];
        try {
          answers = this.hooks.onEvaluate?.(params) ?? [];
        } catch {} // evaluators are additive, never error paths
        return this.respond(id, { ok: true, answers: answers as unknown as Value });
      }
      case "search": {
        try {
          const rows = (await this.hooks.onSearch?.(params)) ?? [];
          return this.respond(id, { ok: true, rows: rows as unknown as Value });
        } catch {
          return this.respond(id, fail("search hook failed"));
        }
      }
      case "hints":
        return this.respond(id, await this.hintsReply(params));
      case "perform":
        return this.respond(id, await this.performReply(params));
      default:
        return this.respond(id, fail(`unknown method: ${method}`));
    }
  }

  private handleInitialize(id: Value, params: Msg) {
    const hostVersion = params.protocol_version;
    if (hostVersion !== PROTOCOL_VERSION) {
      this.respond(id, {
        ok: false,
        protocol_version: PROTOCOL_VERSION,
        error:
          `protocol version mismatch: host v${hostVersion},` +
          ` plugin v${PROTOCOL_VERSION}`,
      });
      this.terminated = true; // terminal: serve() returns, the process exits 0
      return;
    }
    if (this.initialized) {
      return this.respond(id, fail("initialize may only be called once"));
    }
    this.initialized = true;
    this.respond(id, { ok: true, protocol_version: PROTOCOL_VERSION });
    // Fire the async hook AFTER the reply, without blocking the pump.
    void (async () => this.hooks.onStart?.())().catch(() =>
      this.log("error", "start hook failed"),
    );
  }

  private async hintsReply(params: Msg): Promise<Reply> {
    if (!this.hooks.onHints) return { ok: true, targets: [] };
    try {
      const reply = await this.hooks.onHints(params);
      return Array.isArray(reply)
        ? { ok: true, targets: reply }
        : ({ ok: true, ...reply } as Reply); // {targets, context_pid?}
    } catch {
      return fail("hints hook failed");
    }
  }

  private async performReply(params: Msg): Promise<Reply> {
    const kind = params.kind;
    if (!(PERFORM_KINDS as readonly Value[]).includes(kind)) {
      return fail(`unknown perform kind: ${kind}`);
    }
    const hook = {
      resolve: this.hooks.onResolve,
      command: this.hooks.onCommand,
      action: this.hooks.onAction,
      navigate: this.hooks.onNavigate,
    }[kind as (typeof PERFORM_KINDS)[number]];
    if (!hook) return unhandled();
    try {
      return (await hook(params)) ?? ok();
    } catch {
      return fail("perform hook failed"); // mine-but-broke: no fallback
    }
  }

  // ── emitters ───────────────────────────────────────────────────────────

  /** Full-replacement catalog push; each row carries a first-class `source`
   * naming a manifest sources[].name. */
  publish(rows: Row[]) {
    this.notify("publish", { rows: rows as unknown as Value });
  }

  status(segments: { [name: string]: string }) {
    this.notify("status", { segments });
  }

  log(level: string, message: string, fields: { [key: string]: Value } = {}) {
    this.notify("log", { level, message, fields });
  }

  // ── serve loop ─────────────────────────────────────────────────────────

  /** Reads frames until stdin EOF — the shutdown signal: onShutdown runs,
   * serve() resolves, the process exits 0. Oversized lines are discarded
   * (never buffered whole); the stream self-heals at the next newline. */
  async serve() {
    const utf8 = new TextDecoder();
    for await (const chunk of Bun.stdin.stream()) {
      this.buf += utf8.decode(chunk, { stream: true });
      let nl: number;
      while ((nl = this.buf.indexOf("\n")) >= 0) {
        const line = this.buf.slice(0, nl);
        this.buf = this.buf.slice(nl + 1);
        if (this.skipping) {
          this.skipping = false; // tail of an oversized line
          continue;
        }
        if (Buffer.byteLength(line) > MAX_FRAME_BYTES) continue;
        this.dispatch(line);
        if (this.terminated) return; // version mismatch: exit without cleanup
      }
      if (this.done) break; // stdout died: wind down through cleanup
      // UTF-16 units under-count UTF-8 bytes, so this only bounds memory;
      // the per-line byte check above is the precise gate.
      if (this.skipping) this.buf = "";
      else if (this.buf.length > MAX_FRAME_BYTES) {
        this.buf = "";
        this.skipping = true;
      }
    }
    if (this.buf.trim() && !this.skipping) this.dispatch(this.buf); // EOF tail
    if (this.terminated) return;
    this.handleEof();
    try {
      await this.hooks.onShutdown?.();
    } catch {} // exiting anyway; stdout may already be gone
  }
}
